src/puppetlabs/puppetdb/mq_listener.clj

```diff
@@ -133,174 +156,104 @@
-        on-discard     #(handle-command-discard % discard)
-        on-parse-error #(handle-parse-error %1 %2 discard)
-        on-fatal       #(handle-command-failure %1 %2 discard)
-        on-retry       #(handle-command-retry %1 %2 publish)]
-    (-> message-fn
-        (wrap-with-discard on-discard maximum-allowable-retries)
-        (wrap-with-exception-handling on-retry on-fatal)
-        (wrap-with-command-parser on-parse-error)
-        (wrap-with-meter (global-metric :seen))
-        (wrap-with-thread-name "command-proc"))))
+(defn mark-both-metrics!
+  "Calls `mark!` on the global and command specific metric for `k`"
+  [command version k]
+  (mark! (global-metric k))
+  (mark! (cmd-metric command version k)))
+
+(defn update-both-metrics!
+  "Calls `update!` on the global and command specific metric for `k`"
+  [command version k v]
+  (update! (global-metric k) v)
+  (update! (cmd-metric command version k) v))
+
+(defn call-with-command-metrics
+  "Invokes `f` including the related metrics updates"
+  [command version retries f]
+  (create-metrics-for-command! command version)
+
+  (mark-both-metrics! command version :seen)
+  (update-both-metrics! command version :retry-counts retries)
+
+  (mutils/multitime!
+   [(global-metric :processing-time)
+    (cmd-metric command version :processing-time)]
+
+   (let [command-result (f)]
+     (mark-both-metrics! command version :processed)
+     command-result)))
+
+;; Primarily a test hook
+(defn discard-message [message q dlo]
+  (dlo/discard-cmdref message q dlo))
+
+(defn message-handler
+  "Processes the message via (process-message msg), retrying messages
+  that fail via (delay-message msg), and discarding messages that have
+  fatal errors or have exceeded their maximum allowed attempts."
+  [q dlo delay-message process-message]
+  (fn [{:keys [certname command version delete? id] :as cmdref}]
+    (try+
+     ;; If the message is a delete?, there's no need to parse it
+     ;; below, it's only going to be removed
+     (if delete?
+       (do
+         (process-message cmdref)
+         (queue/ack-command q {:entry (queue/cmdref->entry cmdref)})
+         (update-counter! :depth command version dec!)
+         (update-counter! :invalidated command version dec!))
+       (let [cmd (queue/cmdref->cmd q cmdref)
+             retries (count (:attempts cmdref))]
+         (try+
+          (call-with-command-metrics command version retries
+                                     #(process-message cmd))
+          (queue/ack-command q cmd)
+          (update-counter! :depth command version dec!)
+
+          (catch fatal? obj
+            (mark! (global-metric :fatal))
+            (let [ex (:cause obj)]
+              (log/error
+               (:wrapper &throw-context)
+               (i18n/trs "[{0}] [{1}] Fatal error on attempt {2} for {3}" id command retries certname))
+              (-> cmd
+                  (queue/cons-attempt ex)
+                  (discard-message q dlo))))
```

Comment:
If I'm reading this right, we have some cases in this function where we only update the metrics after the message is out of the queue, and others where we do the reverse; behavior should only differ if there's an unexpected exception, but I thought I'd mention it.